### PR TITLE
Fix race-condition in multi-threaded analysis

### PR DESCRIPTION
**Pull request by uklotzde**
https://mixxx.zulipchat.com/#narrow/stream/109695-_support/topic/Bulk.20Analysing.20seeming.20to.20freeze

This fixes a race condition in the multi-threaded analysis where some worker threads might get stuck forever by not resuming after they received a next track for analysis.

The public method was lacking a mutex lock!

I found this by adding more detailed log messages. Those changes could later be merged to master.

### CODE REVIEW

**Comment by uklotzde** [comment]
Ready. Worker thread needs to be woken up unconditionally to avoid race conditions.

**Comment by uklotzde** [comment]
Force pushed a type fix in comment

**Comment by daschuer** [comment]
Unfortunately, I am not able to wrap my head around it in a short time. 

It might work now, but I would prefer a fix that can be easily followed.  
The source of confusion it the the combined function of m_suspend. 
It transports the suspend and the wake command and holds the state in the same time. 
I think it would be better two have two or three values, one for each function.

Do you think it will work to improve the situation. I am unsure. 
If you like I can poke around to verify my idea.

**Comment by uklotzde** [comment]
This is easiest fix I can come up with.

m_suspend is just a simple flag. The only issue was the race condition between notify_on() and wait().

1. Worker thread invokes m_sleepWaitCond.wait()
2. Other thread invokes m_sleepWaitCond.notify_one(), but no thread is waiting
3. Worker thread blocks on m_sleepWaitCond forever and doesn't receive the notification

This is now documented in WorkerThread::wake().

**Comment by uklotzde** [comment]
The logs clearly showed that one thread got assigned a track for analysis but then never started the analysis and got stuck.

**Comment by uklotzde** [comment]
If anyone has a better idea, feel free to rewrite everything. I hope I don't need to touch this again, just fixing my own mistakes.

**Comment by uklotzde** [comment]
The compare_exchange_strong() became obsolete and can be replaced by simply calling store(). The current value doesn't matter and the worker thread only reads this value.

**Comment by Be-ing** [comment]
This is an annoying bug, but no need to hold back the beta release for it.

**Review by daschuer — CHANGES_REQUESTED** [review]
It took me a while to understand that there are two types of suspend. 
On is an external suspend call, the other is the FetchWorkResult::Suspend.
It would be nice to not call the suspend() function from intern and set the flag directly to make it obvious that there are two control flows setting the suspend flag().  

The next source of confusion is the overloaded sleepWhileSuspended() function, 
I think it would helpful to give them different names. something like: 
sleepIfSuspended() 
sleepNow(); 
or such. 

I assume that the same thread is in charge to call suspend()  an resume() from external. 

The internal suspend() call destroys this deterministic. 

Let's assume a FetchWorkResult::Suspend is scheduled as a final task.
Than the scheduler receives new tasks and calls resume() this sets immediately suspend to falls (was already false) than it calls wake() and sleeps until the FetchWorkResult::Suspend is found. 
This calls suspend() and sets suspend to true, falls asleap, and is waked by the waiting wake call. 
It falls immediately asleep again because suspend is still true. 
FetchWorkResult::Suspend wins. 
If FetchWorkResult::Suspend is found a bit earlier, resume () wins. 

Who should win?

**Comment by daschuer** [comment]
Only WorkerThreadScheduler makes use of FetchWorkResult::Suspend;
Can we dispose it and use FetchWorkResult::Idle istead? 
Than the only one thread has write access to m_suspend.

File: src/util/workerthread.cpp
Comment (uklotzde) [reply src/util/workerthread.cpp:97]: The mutex is only  locked for a very short time.

**Comment by uklotzde** [comment]
I have removed all the unused and potentially dangerous result options from tryFetchWorkItems()/FetchWorkResult.

File: src/util/workerthread.cpp
Comment (daschuer) [reply src/util/workerthread.cpp:103]: Ah, now I see, I was concerned about this lock 
https://github.com/mixxxdj/mixxx/blob/d0ca0e55b20b784d67fa7883ce580c16edb5e529/src/util/workerthread.cpp#L141

**Review by daschuer — APPROVED** [review]

**Comment by daschuer** [comment]
Much better, Thank you. LGTM, waiting for CI.

**Pull request merged**
